### ISSUE TITLE
Add AD-preserving trailing-batch gather/index-select for batched contractions

### ISSUE THREAD

**Issue by shinaoka** [tensor4all/tenferro-rs#856]
## Status

This is the canonical follow-up to #544 for TreeTN-style batched contractions.

#544 is already closed/completed and covers the general torch-style shape-packing surface:

- `unsqueeze`
- `squeeze`
- `stack`
- `cat`

This issue should **not** reimplement that stack/cat work. Instead, it should build on #544 and add the missing pieces needed by downstream `tensor4all-rs` TreeTN cached batch evaluation:

- use `stack(..., dim = -1)` as the standard rightmost-batch packing path;
- expose gather/index-select along an arbitrary axis, especially `axis = -1`;
- preserve AD through gather with scatter-add reverse semantics;
- allocate through the normal context/backend buffer path and avoid unnecessary zero-fill;
- verify the produced operands work efficiently with retained-batch contraction.

## Goal

Add a high-level, AD-preserving way to index batched tensor operands for contractions, with the batch axis placed on the **rightmost/trailing dimension**.

TreeTN cached batch evaluation needs to express this without materializing tensors into Rust-side scalar buffers:

1. stack per-assignment tensors into one rightmost-batched tensor using the #544 stack API;
2. gather/select entries along the rightmost batch axis to align cached messages with parent assignments;
3. contract all operands in one retained-batch contraction;
4. preserve no-grad, forward-mode, and reverse-mode behavior through the whole path.

The current tensor4all-side workaround stacks/gathers values by materializing tensors into scalar buffers and reconstructing dense tensors. That is not acceptable as the final design because it:

- breaks or bypasses tenferro AD graph/tape semantics;
- adds avoidable host-side copies and dtype dispatch;
- creates tensor4all-local shape/indexing code that belongs in tenferro;
- can put the batch axis in a layout that is not canonical for tenferro batched GEMM;
- bypasses context/backend buffer-pool allocation opportunities.

## Context

Related downstream work:

- `tensor4all/tensor4all-rs#503` uses TreeTN cached batch evaluation and needs this issue before the stack/gather helpers can become the final AD-compatible implementation.

Relevant tenferro implementation points to inspect first:

- `tenferro/src/compiler/mod.rs`: `DotGeneral` canonicalization already treats batch dimensions as trailing dimensions.
- `tenferro-tensor/src/cpu/gemm/mod.rs`: CPU batched GEMM layout code assumes batch dimensions are trailing/rightmost in canonical form.
- `tenferro-ops/src/ad/indexing.rs`: existing Gather/Scatter AD rules likely provide the right substrate for reverse-mode gather via scatter-add.
- `tenferro-ops/src/ext_op.rs`: ExtensionOp exists as an escape hatch, but this issue should prefer first-class standard tensor/indexing APIs because the operation is generic, not TreeTN-specific.

### Rightmost batch layout

TreeTN should represent the assignment/batch axis as the last dimension.

Example stack convention using #544:

```text
input tensors:       [i, j, k]
N assignments:       N
stacked output:      [i, j, k, N]
stack axis:          -1 / rank
```

This issue does not need a new stack implementation if #544 already provides `Tensor::stack(..., dim)`. It does need tests that make `dim = -1` a first-class path because this is the canonical layout for downstream batched contraction.

### Gather / index-select along one axis

TreeTN needs to gather cached messages along the assignment axis.

Example shape convention:

```text
source message:      [bond_dims..., source_assignments]
positions:           [target_assignments]   // host-known positions are enough for TreeTN
output message:      [bond_dims..., target_assignments]
gather axis:         -1
```

Minimum API needed by tensor4all:

```rust
Tensor::index_select(&self, axis: isize, positions: &[usize]) -> Result<Tensor>
```

or an equivalent public API name such as `take_axis` / `gather_axis` if it better matches tenferro conventions.

Requirements:

- `axis` supports negative indexing, especially `axis = -1`.
- Repeated positions are valid.
- Reverse-mode pullback is scatter-add into the source axis, not overwrite. For example, if `positions = [2, 2, 5]`, the cotangents for output slots 0 and 1 both contribute to source slot 2.
- Forward-mode tangent is gathered with the same positions.
- Positions are not differentiable.
- Shape errors are typed and explicit: invalid axis, out-of-range position, unsupported layout/device, empty positions if rejected.

## Buffer allocation and initialization requirements

Forward stack/gather outputs are fully overwritten. The implementation should therefore avoid unnecessary allocation and initialization overhead.

Concrete requirements:

- output tensors are allocated through the normal tenferro context/backend allocation path so buffer-pool reuse applies;
- trailing-axis stack, if touched here, allocates one output buffer and fills each slice directly;
- `index_select` / `gather_axis` allocates one output buffer and fills every selected slice directly;
- forward outputs must not be zero-initialized before being fully overwritten unless the backend abstraction strictly requires it;
- zero initialization is appropriate only where required by semantics, for example reverse-mode scatter-add accumulation into a cotangent buffer;
- benchmarks or tests should catch accidental scalar materialization, unnecessary intermediate buffers, or zero-fill-before-overwrite behavior where practical.

### 1. Audit existing indexing substrate

Before implementing a new primitive, inspect current support for:

- `StdTensorOp::Gather` / `StdTensorOp::Scatter`;
- eager backend gather/scatter kernels;
- AD rules in `tenferro-ops/src/ad/indexing.rs`;
- dynamic `tenferro::Tensor` wrappers for gather/scatter/select/narrow;
- allocation APIs that use the context/backend buffer pool without forcing zero initialization.

Prefer exposing or lightly wrapping existing standard ops over introducing a new ExtensionOp.

### 2. Add tests first

Suggested tests:

- `Tensor::stack(..., dim = -1)` from #544 packs rank-0, rank-1, and rank-2 dense tensors into the expected trailing-batch layout;
- trailing-axis stack preserves reverse-mode gradients by splitting cotangents back to inputs;
- `index_select(axis = -1, positions)` returns expected primal values for a column-major dense tensor;
- `index_select` supports repeated positions;
- repeated-position reverse-mode gradient uses scatter-add;
- forward-mode tangent follows the same gather positions;
- invalid axis and out-of-range position return typed errors;
- no scalar extraction path is used for stack/gather outputs.

A small numerical reverse-mode test should make duplicate accumulation impossible to miss:

```text
x shape:      [3]
positions:    [1, 1, 2]
y = index_select(x, 0, positions)
loss = sum(weights * y), weights = [10, 20, 30]
expected dx = [0, 30, 30]
```

### 3. Add typed/eager API

At the `tenferro-tensor` level, add or expose the low-level operation needed to copy strided slices along one axis into a dense output.

The implementation should:

- allocate one output tensor through the normal backend/context path;
- fill selected slice windows in column-major order;
- avoid building intermediate scalar buffers;
- handle `axis = -1` efficiently because this is the main TreeTN path;
- reject unsupported memory spaces/layouts with explicit typed errors rather than silently transferring or flattening.

### 4. Add dynamic `tenferro::Tensor` wrapper and AD

Expose a public dynamic API on `tenferro::Tensor`.

Forward behavior:

- primal: gather/index-select the primal payload;
- tangent: gather/index-select the tangent payload when present.

Reverse behavior:

- register a pullback that scatters the cotangent into a zero-like source cotangent with add accumulation at repeated positions.

If existing Gather/Scatter AD rules already provide this exactly, the wrapper should lower to those standard ops instead of implementing custom AD logic.

### 5. Verify retained-batch contraction shape

Add a small integration-style test demonstrating the intended layout:

```text
A: [m, k, b]
B: [k, n, b]
contract over k, retain b
C: [m, n, b]
```

This should exercise the same rightmost batch axis that TreeTN will use. The test does not need to depend on tensor4all.

### 6. Benchmark / regression guard

Add a focused benchmark or lightweight perf regression check if the repo already has a suitable benchmark harness.

The benchmark should help catch:

- scalar-materialization fallback;
- extra host buffers;
- zero-fill-before-overwrite behavior;
- avoidable transpose/copy caused by not using the trailing batch layout;
- retained-batch contraction regressions when using the produced operands.

The goal is not to guarantee a specific speedup in this issue, but to prevent pathological overhead in the TreeTN path.

## Acceptance criteria

- #544 remains the source of truth for general `stack` / `cat`; this issue does not duplicate that implementation.
- A public API exists for selecting/gathering along an arbitrary axis, including `axis = -1`.
- Repeated gather positions accumulate gradients correctly in reverse mode.
- Forward-mode and no-grad paths both produce correct primal/tangent values.
- TreeTN-style trailing batch layout is covered by tests.
- The implementation does not materialize tensors through Rust-side scalar extraction.
- Output allocation uses the normal context/backend buffer path where available.
- Fully overwritten forward outputs avoid unnecessary zero initialization.
- The implementation either reuses existing Gather/Scatter standard ops or clearly documents why a new op was needed.
- The API and rustdoc examples make it clear how to build rightmost-batch operands for batched contractions.

## Out of scope

- Tensor4all-side TreeTN changes. Those should be done after this tenferro API exists.
- Reimplementing #544 stack/cat.
- A TreeTN-specific ExtensionOp. ExtensionOp remains a fallback only if first-class standard ops cannot support this cleanly.
- Full NumPy-style advanced indexing or broadcasting index tensors, unless it falls out naturally from existing Gather support.
- GPU-native implementation, unless the current backend already has the necessary primitives. A typed unsupported error is acceptable for phase 1.
- Structured/sparse/diagonal layout support for stack/gather in phase 1.

**Comment by shinaoka** [comment]
Related prerequisite/complement: #544.

This issue should treat #544 as the general stack/cat API work. If #544 lands `Tensor::stack(..., dim)`, this issue should not duplicate that implementation; it should focus on:

- requiring and testing the trailing-axis `dim = -1` path for batched contractions;
- adding or exposing the gather/index-select API needed to align cached batch messages;
- verifying scatter-add reverse-mode behavior for repeated gather positions;
- documenting the rightmost-batch layout expected by tenferro batched GEMM.

**Comment by shinaoka** [comment]
Additional implementation requirement: use the context/backend buffer-pool allocation path and avoid unnecessary initialization.

For TreeTN batched evaluation this is not only a performance detail; it is part of the intended API contract for these operations. `stack` / `index_select` / `gather_axis` should not allocate ad hoc host buffers when a tenferro context/backend buffer pool can provide the output storage, and they should not zero-initialize output buffers that will be fully overwritten by slice copies or gather writes.

Concrete expectations for this issue:

- output tensors should be allocated through the normal tenferro context/backend allocation path, so existing buffer-pool reuse applies;
- dense `stack(..., dim = -1)` should allocate one output buffer and fill each slice directly;
- `index_select` / `gather_axis` should allocate one output buffer and fill every selected slice directly;
- zero-initialization is only appropriate when required by semantics, e.g. reverse-mode scatter-add accumulation into a cotangent buffer, not for fully overwritten forward outputs;
- tests or benchmarks should be able to catch accidental extra materialization, intermediate scalar buffers, or zero-fill-before-overwrite behavior where practical.

**Comment by shinaoka** [comment]
Completed by #857.

The merged PR adds concrete, eager, and traced `index_select` / `stack` APIs for the trailing-batch shape-packing path, lowers `index_select` through existing `Gather` semantics, keeps repeated-position AD accumulation covered by tests, and verifies the retained-batch contraction layout.
